File: src/architecture.erl
%%%-------------------------------------------------------------------
%%% @author borja
%%% @doc
%%%
%%% @end
%%%-------------------------------------------------------------------
-module(architecture).
-compile([export_all, nowarn_export_all]). %%TODO: To delete after build

%% API
-export([fields/1, load/1]).
-export_type([id/0, architecture/0]).

-type id() :: {Name :: atom(), architecture}.
-define(ARCHITECTURE_ID(Name), {Name, architecture}).

-record(architecture, {
    id   :: id(),
    dim  :: [LayerSize :: integer()],
    type :: model:nature()
}).
-type architecture() :: #architecture{}.

-define(DEF_DIM,  [3]).
-define(DEF_TYPE, recurrent).
-define(MAYBE(Probability), Probability > rand:uniform() andalso).


%%%===================================================================
%%% API
%%%===================================================================

%%-------------------------------------------------------------------
%% @doc Returns the record fields of an element.
%% @end
%%-------------------------------------------------------------------
-spec fields(Atom :: architecture) -> ListOfFields :: [atom()].
fields(architecture) -> record_info(fields, architecture).

%%-------------------------------------------------------------------
%% @doc Loads the architectures from the indicated modules. 
%% @end
%%-------------------------------------------------------------------
-spec load(Modules :: [Module :: module()]) -> ok.
load(Modules) -> 
    Sets = set_modules(Modules, sets:new()), 
    edb:write(sets:to_list(Sets)),
    ok. 

%%-------------------------------------------------------------------
%% @doc Generates a network id from a defined architecture. 
%% @end
%%-------------------------------------------------------------------
-spec new_network(Architecture :: architecture()) -> 
    Network_id :: enn:id().
new_network(A) -> 
    Nature = A#architecture.type,
    enn:compile(
        model:Nature([layer:dense(N) || N <- A#architecture.dim])
    ).

%%-------------------------------------------------------------------
%% @doc Mutates a network returning the new mutation id. 
%% @end
%%-------------------------------------------------------------------
-spec mutate(Network_id :: enn:id()) -> 
    NewNetwork_id :: enn:id().
mutate(_Network_id) -> 
    error(undefined). % TODO


%%%===================================================================
%%% Network transformations 
%%%===================================================================


%%%===================================================================
%%% Links transformations 
%%%===================================================================


%%%===================================================================
%%% Neuron transformations 
%%%===================================================================


%%====================================================================
%% Internal functions
%%====================================================================

% --------------------------------------------------------------------
set_modules([Module | Modules], Sets) ->
    set_modules(Modules,
        add_architectures(Module:all(), Module, Sets));
set_modules([], Sets) -> 
    Sets.

add_architectures([Name | Architectures], Module, Sets) -> 
    Definition   = Module:Name(),
    Architecture = #architecture{
        id   = ?ARCHITECTURE_ID(Name), 
        dim  = maps:get(initial_dimensions, Definition,  ?DEF_DIM),
        type = maps:get(        model_type, Definition, ?DEF_TYPE)
    },
    add_architectures(Architectures, Module, 
        sets:add_element(Architecture, Sets));
add_architectures([], _, Sets) -> 
    Sets.


%%====================================================================
%% Eunit white box tests
%%====================================================================

% --------------------------------------------------------------------
% TESTS DESCRIPTIONS -------------------------------------------------

% --------------------------------------------------------------------
% SPECIFIC SETUP FUNCTIONS -------------------------------------------

% --------------------------------------------------------------------
% ACTUAL TESTS -------------------------------------------------------

% --------------------------------------------------------------------
% SPECIFIC HELPER FUNCTIONS ------------------------------------------

File: lib/architecture.ex
defmodule Architecture do
  @moduledoc """
  """
  defstruct id: nil, dim: nil, type: nil

  @doc false
  defmacro __using__(_opts) do
    quote do
      import Architecture

      # Initialize @architectures to an empty list
      @architectures []

      # Invoke Architecture.__before_compile__/1 before the module is compiled
      @before_compile Architecture
    end
  end

  @doc false
  defmacro __before_compile__(_env) do
    quote do
      # Creates a function to return the architectures
      def architectures, do: @architectures
    end
  end

  @doc """
  Defines a new architecture
  """
  defmacro defarchitecture(name, do: block) do
    architecture_name = String.to_atom(name)
    var_architecture = Macro.var(:architecture, nil)

    def_architecture =
      Macro.escape(%Architecture{
        id: Database.id(architecture_name, :architecture)
      })

    quote do
      @architectures [unquote(architecture_name) | @architectures]
      @spec unquote(architecture_name)() :: Architecture.t()
      def unquote(architecture_name)() do
        unquote(var_architecture) = unquote(def_architecture)
        unquote(block)
        unquote(var_architecture)
      end
    end
  end

  @doc """
  Defines the dimensions for an architecture. Should be placed inside
  a defarchitecture ... end instance
  """
  defmacro defdimensions(dimensions) do
    quote do
      var!(architecture) = Map.put(var!(architecture), :dim, unquote(dimensions))
    end
  end

  @doc """
  Defines the model type for an architecture. Should be placed inside
  a defarchitecture ... end instance
  """
  defmacro defmodel_type(model_type) do
    quote do
      var!(architecture) = Map.put(var!(architecture), :type, unquote(model_type))
    end
  end

  ### =================================================================
  ###  API
  ### =================================================================

  @doc """
  Returns the fields of an architecture structure.
  Should run inside :mnesia transaction (or Database.run/1)
  """
  @spec fields() :: [atom()]
  def fields(), do: Map.from_struct(%Architecture{}) |> Map.keys()

  @doc """
  Loads the architectures from the indicated modules.
  Should run inside :mnesia transaction (or Database.run/1)
  """
  @spec load(atom | [atom]) :: any
  def load(modules) when is_list(modules) do
    for m <- modules, do: load(m)
  end

  def load(module) do
    for name <- module.architectures do
      architecture = %Architecture{} = apply(module, name, [])
      Database.write(architecture)
    end
  end

  @doc """
  Mutates the neuronal network
  """
  @spec mutate(:enn.network()) :: :enn.network()
  def mutate(network) do
    {:atomic, :ok} = :enn_edit.transaction(network, &mutate_network/1)
    network
  end

  ### =================================================================
  ###  Internal functions
  ### =================================================================

  # Randomly mutates a neural network -------------------------------
  defp mutate_network(enn) do
    enn
    |> maybe_divide_neuron
    |> maybe_increasse_connections
    |> mutate_neurons
  end

  defp mutate_neurons(enn) do
    neurons = :network.neurons(enn)
    sf = :math.sqrt(Map.get(:network.info(enn), :size))
    Enum.reduce(:ltools.rand(neurons, 1 / sf), enn, &mutate_neuron/2)
  end

  @chance 0.05
  defp maybe_divide_neuron(enn) do
    maybe(@chance, enn, fn enn ->
      neurons = :network.neurons(enn)
      neuron = :ltools.randnth(neurons)
      :enn_edit.divide_neuron(enn, neuron)
    end)
  end

  @chance 0.05
  defp maybe_increasse_connections(enn) do
    maybe(@chance, enn, fn enn ->
      neurons = :network.neurons(enn)
      sf = :math.sqrt(Map.get(:network.info(enn), :size))
      from = :ltools.rand(neurons, 1 / sf)
      to = :ltools.rand(neurons, 1 / sf)
      :enn_edit.connect_allowed(enn, from, to)
    end)
  end

  # Randomly mutates a neuron network -------------------------------
  defp mutate_neuron(neuron, enn) do
    enn
    |> maybe_reinitialise_bias(neuron)
    |> maybe_switch_activation(neuron)
  end

  @chance 0.15
  defp maybe_reinitialise_bias(enn, neuron) do
    maybe(@chance, enn, fn enn ->
      :enn_edit.reinitialise_bias(enn, neuron)
    end)
  end

  @chance 0.15
  @functions [:direct, :sigmoid, :tanh, :elu]
  defp maybe_switch_activation(enn, neuron) do
    maybe(@chance, enn, fn enn ->
      func = :ltools.randnth(@functions)
      :enn_edit.switch_activation(enn, neuron, func)
    end)
  end

  # Maybe applies the function over enn or returns the enn ----------
  defp maybe(probability, enn, fun) do
    case :rand.uniform() do
      x when x < probability -> apply(fun, [enn])
      _ -> enn
    end
  end
end
